File: javascript/packages/linter/src/rules/html-no-block-inside-inline.ts
import { BaseRuleVisitor, isInlineElement, isBlockElement } from "../utils/rule-utils.js"
import { ParserRule } from "../types.js"
import { isHTMLOpenTagNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOpenTagNode, HTMLElementNode, ParseResult } from "@herb-tools/core"

const SVG_HTML_INTEGRATION_POINTS = new Set(["foreignobject", "desc", "title"])

// TODO: refactor using ElementStackVisitor
class BlockInsideInlineVisitor extends BaseRuleVisitor {
  private inlineStack: string[] = []
  private insideSVG = false

  private getElementType(tagName: string): { isInline: boolean; isBlock: boolean; isUnknown: boolean } {
    const isInline = isInlineElement(tagName)
    const isBlock = isBlockElement(tagName)
    const isUnknown = !isInline && !isBlock

    return { isInline, isBlock, isUnknown }
  }

  private addOffenseMessage(tagName: string, isBlock: boolean, openTag: HTMLOpenTagNode): void {
    const parentInline = this.inlineStack[this.inlineStack.length - 1]
    const elementType = isBlock ? "Block-level" : "Unknown"

    this.addOffense(
      `${elementType} element \`<${tagName}>\` cannot be placed inside inline element \`<${parentInline}>\`.`,
      openTag.tag_name!.location,
    )
  }

  private visitInlineElement(node: HTMLElementNode, tagName: string): void {
    this.inlineStack.push(tagName)
    super.visitHTMLElementNode(node)
    this.inlineStack.pop()
  }

  private visitBlockElement(node: HTMLElementNode): void {
    const savedStack = [...this.inlineStack]
    this.inlineStack = []
    super.visitHTMLElementNode(node)
    this.inlineStack = savedStack
  }

  private visitSVGElement(node: HTMLElementNode): void {
    const savedStack = this.inlineStack
    const wasInsideSVG = this.insideSVG

    this.inlineStack = []
    this.insideSVG = true
    super.visitHTMLElementNode(node)
    this.insideSVG = wasInsideSVG
    this.inlineStack = savedStack
  }

  visitHTMLElementNode(node: HTMLElementNode): void {
    if (!isHTMLOpenTagNode(node.open_tag)) {
      super.visitHTMLElementNode(node)
      return
    }

    const tagName = node.open_tag.tag_name?.value.toLowerCase()

    if (!tagName) {
      super.visitHTMLElementNode(node)
      return
    }

    if (this.insideSVG) {
      if (SVG_HTML_INTEGRATION_POINTS.has(tagName)) {
        this.insideSVG = false
        super.visitHTMLElementNode(node)
        this.insideSVG = true
        return
      }

      super.visitHTMLElementNode(node)
      return
    }

    if (tagName === "svg") {
      this.visitSVGElement(node)
      return
    }

    const { isInline, isBlock, isUnknown } = this.getElementType(tagName)

    if ((isBlock || isUnknown) && this.inlineStack.length > 0) {
      this.addOffenseMessage(tagName, isBlock, node.open_tag)
    }

    if (isInline) {
      this.visitInlineElement(node, tagName)
      return
    }

    this.visitBlockElement(node)
  }
}

export class HTMLNoBlockInsideInlineRule extends ParserRule {
  static ruleName = "html-no-block-inside-inline"
  static introducedIn = this.version("0.4.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "error"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new BlockInsideInlineVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_block_inside_inline.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::html_data::{is_block_element, is_inline_element};

use herb::nodes::*;
use herb::union_types::*;
use herb::ParseResult;
use herb::Visitor;
use herb_config::Severity;

pub struct HTMLNoBlockInsideInlineRule;

struct BlockInsideInlineVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  inline_stack: Vec<String>,
}

impl Visitor for BlockInsideInlineVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    let open_tag = match &node.open_tag {
      Some(ERBOpenTagNodeOrHTMLConditionalOpenTagNodeOrHTMLOpenTagNode::HTMLOpenTagNode(tag)) => tag,
      _ => {
        self.walk_html_element_node(node);
        return;
      }
    };

    let tag_name = match &open_tag.tag_name {
      Some(token) => token.value.to_lowercase(),
      None => {
        self.walk_html_element_node(node);
        return;
      }
    };

    let is_inline = is_inline_element(&tag_name);
    let is_block = is_block_element(&tag_name);
    let is_unknown = !is_inline && !is_block;

    if (is_block || is_unknown) && !self.inline_stack.is_empty() {
      let parent_inline = self.inline_stack.last().unwrap().clone();
      let element_type = if is_block { "Block-level" } else { "Unknown" };

      if let Some(ref tag_name_token) = open_tag.tag_name {
        self.offenses.push(UnboundOffense::new(
          self.rule_name,
          format!(
            "{} element `<{}>` cannot be placed inside inline element `<{}>`.",
            element_type, tag_name, parent_inline
          ),
          tag_name_token.location.clone(),
        ));
      }
    }

    if is_inline {
      self.inline_stack.push(tag_name);
      self.walk_html_element_node(node);
      self.inline_stack.pop();
    } else {
      let saved_stack = std::mem::take(&mut self.inline_stack);

      self.walk_html_element_node(node);
      self.inline_stack = saved_stack;
    }
  }
}

impl Rule for HTMLNoBlockInsideInlineRule {
  fn name(&self) -> &'static str {
    "html-no-block-inside-inline"
  }

  fn default_severity(&self) -> Severity {
    Severity::Error
  }

  fn default_enabled(&self) -> bool {
    false
  }
}

impl ParserRule for HTMLNoBlockInsideInlineRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = BlockInsideInlineVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      inline_stack: Vec::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
